keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,330 @@
+  - [Kubelet Configuration](#kubelet-configuration)
+  - [Test Plan](#test-plan)
+- [Graduation Requirements](#graduation-requirements)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Alternatives Considered](#alternatives-considered)
+  - [Other OpenTelemetry Exporters](#other-opentelemetry-exporters)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This Kubernetes Enhancement Proposal (KEP) proposes enhancing the kubelet to allow tracing gRPC and HTTP API requests. 
+It proposes using OpenTelemetry libraries, and exports in the OpenTelemetry format. This is in line with the 
+[API Server enhancement](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/647-apiserver-tracing).
+
+## Motivation
+
+Along with metrics and logs, traces are a useful form of telemetry to aid with debugging incoming requests.
+The kubelet can make use of distributed tracing to improve the ease of use and enable easier analysis of trace data.
+Trace data is structured, providing the detail necessary to debug requests across service boundaries.
+As more core components are instrumented, Kubernetes becomes easier to monitor, manage, and troubleshoot.
+
+### Definitions
+
+**Span**: The smallest unit of a trace.  It has a start and end time. Spans are the building blocks of a trace.
+**Trace**: A collection of Spans which represents work being done by a system. A record of the path of requests through a system.
+**Trace Context**: A reference to a Trace that is designed to be propagated across component boundaries.
+
+### Goals
+
+* The kubelet generates and exports spans for incoming and outgoing requests.
```

Comment:
What is an "incoming request" in the context of the kubelet?

Unless i'm misinterpreting, this should be "generates and exports spans for reconcile-loops it initiates"

Reply:
"kubelet generates and exports spans when interacting with a node's container runtime via Container Runtime Interface and also with the API server" ?

Reply:
Ideally, the kubelet should also generate a span when it initiates work (i.e. a reconciliation loop).  For example, the creation of a pod involves pulling the image, creating the pod sandbox, creating containers, etc.  We want those linked together into a single trace so that we can see what caused the entire reconciliation to be slow.

So in total, it should be something like "generates and exports spans for reconcile loops it initiates and for outgoing requests".

Reply:
rephrased, updated